developer/src/kmc-kmw/src/compiler/messages.ts

```diff
@@ -1,220 +1,47 @@
-import { callbacks, FCompilerWarningsAsErrors } from "./compiler-globals.js";
-
-export let FError = false;
-
-export function ReportError(line: number, msgcode: number, text: string) {  // I1971
-  let flag = CERR_FLAG | CFATAL_FLAG;
-  if(FCompilerWarningsAsErrors) {
-    flag |= CWARN_FLAG;
-  }
-  if (msgcode & flag) {
-    FError = true;
-  }
-  // TODO: these Codes Are All Wrong But We'll  Fix It Soon
-  callbacks.reportMessage({code: msgcode, message: text, line});  // I3310
-}
-
-export const
-  CHINT_FLAG = 0x1000,
-  CWARN_FLAG = 0x2000,
-  CERR_FLAG = 0x4000,
-  CFATAL_FLAG = 0x8000,
-
-// Any messages from the lexical model compiler occupy this range:
-  CERR_LEXICAL_MODEL_MIN                           = 0x0800,
-  CERR_LEXICAL_MODEL_MAX                           = 0x08FF,
```

Comment:
So... are these previously-defined constants no longer accessed?  I'm a bit confused as to how we're eliminating them... and to how the new static functions for generating errors and warnings are assigned their underlying code.  That mechanism isn't clear to me.

As compiler error codes should stay consistent across compiler versions if at all possible, and I'm not clear on how that's happening here, I'm uncomfortable approving at this time.

Reply:
I think these constants were all pulled over from the pascal or C++ and used originally/briefly, but no longer. 

they are assigned manually, and they are scoped to the `CompilerErrorNamespace.KmwCompiler` namespace (numberspace?). 

So for example instead of  `CERR_InvalidBegin                                = 0x400B`  it's now imported as `KmnCompilerMessages.Error_InvalidBegin`  (inherited by KmwCompilerMessages)

which is defined here: https://github.com/keymanapp/keyman/blob/6ac3ccb8cb167aeda07522b177cebcb55ab10452/developer/src/kmc-kmn/src/compiler/messages.ts#L108-L109 

… manually as `SevError | 0x000B`. 

Hope it helps!

Reply:
So yes, I agree that error codes should not change, as far as possible. But I opted to make some changes, for the following reason:

1. The existing codes had a very limited a range of values available (0x000 - 0xFFF)
2. That range was further broken down in to 'namespaces' such as CERR_LEXICAL_MODEL_MIN-MAX and thus the LM compiler only had 256 possible codes available.
3. What's more, different tools reported errors differently, and the LM compiler would throw errors without codes for non-recoverable errors.
4. kmc is a new tool in 17.0, so the need to maintain consistency it and kmcomp is less compelling.
5. Finally, we haven't actually documented the compiler errors anywhere; they were only ever 'documented' in the source (this is not actually a good thing, but it is what it is!)

In terms of the change, the error codes from kmcomp have the same base values as the codes from kmc. However, kmc changes the severity bits (shifted left by 16 bits, and different values) to give us more space for additional codes (severity mask is 0xF00000 instead of 0xF000), adds a namespace mask (0xF000), and puts the values from kmcmplib under the `KmnCompiler` namespace (0x2000). Those are the only differences for the codes.

(Hmm, good to have this discussion because in doing so I just discovered a bug I created in the kmc-kmn compiler messages! see #8998)

Note that the lexical model compiler error code changes have _already_ been approved and merged, way back in #8637. Given there were only two warnings with defined codes, and the remaining errors did not have associated codes (because they threw exceptions), I thought it was better to refactor entirely and cleanup for the future than try to maintain this relationship.

Reply:
See also

https://github.com/keymanapp/keyman/blob/a25f8a723f6724e2a53b1498eb23295f153c39c1/developer/src/kmc-kmn/src/compiler/messages.ts#L81-L93